lib/oelite/magiccache.py

```diff
@@ -0,0 +1,46 @@
+import magic
+
+class OEliteMagic:
+
+    def __init__(self, flags, filename):
+        self.refs = 1
+        self.magic  = magic.open(flags)
+        self.filename = filename
+        self.magic.load(self.filename)
+
+    def get(self):
+        self.refs += 1
+        return self
+
+    def file(self, filename):
+        return self.magic.file(filename)
+
+    def descriptor(self, fd):
+        return self.magic.descriptor(fd)
+
+    def close(self):
+        self.refs -= 1
+        if self.refs == 0:
+            self.magic.close()
+            self.magic = None
+
+
+cache = dict()
+
+def open(flags = magic.MAGIC_NONE, filename=None):
+    t = (flags,filename)
+    if t not in cache:
+        cache[t] = OEliteMagic(flags, filename)
+    return cache[t].get()
```

Comment:
Yes. The cache owns one reference (the initial one created by the constructor), the caller obtains another. When the caller is done, he calls close(), which decrements refs, but the object is kept alive precisely by the reference from the cache dict. If somebody calls clear_cache(), the cache dict drops all its references, but any OEliteMagic in active use is kept alive by the reference that was handed out. 

This is of course all very academic since we're only ever going to have one element in the cache, and we're never calling clear_cache (I've done it for debugging), but it felt cleaner to do it this way rather than hardcoding the singleton.